File: backend/app/models/asset_disposition.py
from datetime import datetime
from decimal import Decimal

from sqlalchemy import DateTime, ForeignKey, Numeric, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class AssetDisposition(Base):
    __tablename__ = "asset_dispositions"

    id: Mapped[int] = mapped_column(
        primary_key=True,
        autoincrement=True,
    )

    asset_id: Mapped[int] = mapped_column(
        ForeignKey("assets.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )

    disposition_type: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
        index=True,
    )

    disposition_status: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
        default="pending",
        server_default="pending",
        index=True,
    )

    disposition_date: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    processed_by: Mapped[str | None] = mapped_column(
        String(150),
        nullable=True,
    )

    disposition_reference: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
        index=True,
    )

    recipient_name: Mapped[str | None] = mapped_column(
        String(150),
        nullable=True,
    )

    recipient_reference: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
    )

    amount: Mapped[Decimal | None] = mapped_column(
        Numeric(12, 2),
        nullable=True,
    )

    currency: Mapped[str | None] = mapped_column(
        String(3),
        nullable=True,
    )

    notes: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

File: frontend/src/types/assetDisposition.ts
export type DispositionType =
  | 'resale'
  | 'reuse'
  | 'internal_use'
  | 'recycling'
  | 'scrap'
  | 'returned'
  | 'donated'
  | 'other'

export type DispositionStatus =
  | 'pending'
  | 'approved'
  | 'completed'
  | 'cancelled'

export type AssetDisposition = {
  id: number
  asset_id: number
  disposition_type: DispositionType
  disposition_status: DispositionStatus
  disposition_date: string
  processed_by: string | null
  disposition_reference: string | null
  recipient_name: string | null
  recipient_reference: string | null
  amount: string | null
  currency: string | null
  notes: string | null
  created_at: string
  updated_at: string
}